### PR TITLE
Closed wire cut

### LINKED ISSUE #56: `chamfer` and `bevel` not accepted on wire extremities
Currently, a `bevel()` works on any double-sided points of a `Web` and a `Wire`. However Wire's extremities might be considered double-sided points as well when `Wire.closed = True` and `bevel()` currently does not consider that.

```python
from madcad import *

corner = 0   # index 0 in the wire is not accepted, nor -1,  any other place is fine

outline = regon(Axis(O,Z), 1, 5)
bevel(outline, [corner], ('width', 0.3))

show([ outline ])
```
```python
Traceback (most recent call last):
  File "/tmp/test.py", line 4, in <module>
    bevel(outline, [0], ('width', 0.3))
  File "/usr/lib/python3.8/functools.py", line 875, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
  File "/home/jimy/.local/lib/python3.8/site-packages/madcad/cut.py", line 939, in wire_bevel
    cuts = set(wire_multicut(wire, points, cutter))
  File "/home/jimy/.local/lib/python3.8/site-packages/madcad/cut.py", line 890, in wire_multicut
    raise MeshError('a chamfer cannot have only one side')
madcad.mesh.container.MeshError: a chamfer cannot have only one side
```

with `corner = 1` we get the expected behavior
image

### PR DESCRIPTION
**Pull request by GlennWSo**
resolves #56 

Made changes so chamfer and bevels can be added to all corners of closed wires.

Along the way i noted the cutter_distance was incorrect.

## Picture of bevels on all corners
closed_bevel_tube

### CODE REVIEW

**Comment by jimy-byerley** [comment]
Nice !
At the first sight I have few remarks

- you removed the propagations from the cutted points
  https://github.com/jimy-byerley/pymadcad/blob/10ee7849bba940f08fe0cb82bbccec69c8ad7617/madcad/cut.py#L909
these propagations guaranteed that when chamfering points on or between curves, the chamfer was looking along the curve (propagating edge by edge) to find the proper intersection place

  You can try it with the following:
  ```python
  P0 = vec3(0.08051, -2.024e-08, 0.1698)
  w = wire([
	  ArcThrough(
		  vec3(-0.1988, 1.5e-10, -0.001258),
		  vec3(-0.06793, -1.35e-08, 0.1132),
		  P0),
	  ArcThrough(
		  P0,
		  vec3(0.08932, -3.899e-09, 0.03271),
		  vec3(0.2327, 1.335e-08, -0.112)),
	  ])
  bevel(w, [4], ('width', 0.15))
  ```
  expected result (with the current master):
  
  image


- you make use of `list.index()` instead of iterating with an index
  https://github.com/jimy-byerley/pymadcad/blob/10ee7849bba940f08fe0cb82bbccec69c8ad7617/madcad/cut.py#L921) 
  `index()` is a `O(n)` long search thus increasing a lot the computation time when called repeatedly. This is why the original implementation used indices in the buffer elements instead of iterating over the buffer's elements and then searching for their index. We should keep it that way to stay as fast as possible.

**Comment by GlennWSo** [comment]
okey, will give it another go :)

**Comment by GlennWSo** [comment]
im passing the tests now, i think :)

But i still have more TODO. Mainly removing index() calls from wire_bevel.
Ill write again when i think the changes are ready for review

cut_curved_wire
closed_bevel

File: madcad/cut.py [hunk 1]
```diff
@@ -931,7 +954,7 @@ def wire_multicut(wire, points, cutter):
 def wire_chamfer(wire, points, cutter):
 	wire_multicut(wire, points, cutter)
 
-@bevel.register(Wire)
+@bevel.register(Wire)  # TODO replace index() 
 def wire_bevel(wire, points, cutter, resolution=None):
```
Comment (GlennWSo) [inline madcad/cut.py:958]: This method seems to work, but it still uses expensive index() calls

**Comment by jimy-byerley** [comment]
I added calls to `Wire.check()` that tests the wire indices and consistency, this is quite usefull for non-regression tests

**Pull request merged**